compiler/qsc_circuit/src/builder.rs

```diff
@@ -206,7 +206,13 @@ impl Backend for Builder {
                 Some(classical_args)
             },
         ));
-        Some(Ok(Value::unit()))
+
+        match name {
+            // Special case this known intrinsic to match the simulator
+            // behavior, so that our samples will work
+            "BeginEstimateCaching" => Some(Ok(Value::Bool(true))),
```

Comment:
If the custom intrinsic is non-Unit and it is not `BeginEstimateCaching`, it seems like building the circuit might break some time later because the evaluator could get `Unit` when in reality it is expecting another kind of value.

If this is the case, I don't think this needs to be addressed in this PR since custom instrinsics outside the standard library are rare, but we should leave a comment or track it if we plan to address it.

Reply:
I think the behavior could be similar to the evaluator one:
```rust
"BeginEstimateCaching" => Some(Ok(Value::Bool(true))),
            "EndEstimateCaching"
            | "AccountForEstimatesInternal"
            | "BeginRepeatEstimatesInternal"
            | "EndRepeatEstimatesInternal" => Some(Ok(Value::unit())),
            _ => Some(Err("Custom intrinsic operations, such as '{name}' are not supported for circuit visualization")),
```

This way we and the user would know an unsupported scenario occurred instead of a seemingly random evaluation error.

Reply:
But then I think we're erring on the side of *too* restrictive. Custom intrinsics are okay to render in circuits, as long as they don't return a value that can manipulate classical control flow. 

One problem here is that we just don't know the intended return type of the custom intrinsic (because we don't have access to the compilation). If we did, we could narrow down the error case to "non-unit returning custom intrinsics are not supported").

Custom intrinsics that return anything other than `Unit` will fail Base Profile validation (`Qsc.BaseProfCk.UnsupportedIntrinsic`), so that helps a little bit. If you're using `Unrestricted`, then you're probably targeting the simulator and not implementing your own custom intrinsic.

Maybe we just need a circuits-specific pass to detect unsupported custom intrinsics _before_ we get to the evaluator. Once we're here, we just don't have enough information...

Agree that this should probably be a separate issue/PR.

Reply:
I think the checks in base profile and the fact that they wouldn't be supported for simulation if someone is using `Unrestricted` seem enough reason to probably not worry about this.

Sorry for the churn!

Reply:
np, thanks for helping me think through the scenario.